sqlstore/src/main/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateSampleDao.java

```diff
@@ -71,22 +77,50 @@
   @Autowired
   private JdbcTemplate template;
 
+  @Transactional(propagation = Propagation.MANDATORY)
   @Override
-  public Long addSample(Sample sample) throws IOException, MisoNamingException {
+  public Long addSample(final Sample sample) throws IOException, MisoNamingException {
     Date now = new Date();
     sample.setLastUpdated(now);
-    // if the sample naming scheme doesn't allow duplicates, and a sample alias already exists
-    if (!namingScheme.allowDuplicateEntityNameFor("alias") && !listByAlias(sample.getAlias()).isEmpty()) {
-      throw new IOException("NEW: A sample with this alias already exists in the database");
-    }
-    if (sample.getSecurityProfile() != null) {
-      sample.setSecurityProfileId(getSecurityProfileDao().save(sample.getSecurityProfile()));
-    }
-    // We can't generate the name until we have the ID and we don't have an ID until we start the persistance. So, we assign a temporary
+
+    // We can't generate the name until we have the ID and we don't have an ID until we start the persistence. So, we assign a temporary
     // name.
-    sample.setName("TEMPORARY_S" + sample.hashCode());
+    sample.setName(generateTemporaryName());
     long id = (Long) currentSession().save(sample);
+
+    sessionFactory.getCurrentSession().doWork(new Work() {
+
+      @Override
+      public void execute(Connection connection) throws SQLException {
+        try {
+          if (!namingScheme.allowDuplicateEntityNameFor("alias") && aliasExists(sample.getAlias())) {
+            // Size is greater than 1, since we've just added this to the db under a temporary name.
+            throw new IOException(String.format("NEW: A sample with this alias '%s' already exists in the database", sample.getAlias()));
+          }
+        } catch (IOException e) {
+          throw new SQLException(e);
+        }
+
+      }
+    });
+
+    if (sample.getSecurityProfile() != null) {
+      sessionFactory.getCurrentSession().doWork(new Work() {
+
+        @Override
+        public void execute(Connection connection) throws SQLException {
```

Comment:
The `sessionFactory.getCurrentSession().doWork(new Work() {` sections allow JDBC calls to run without completing the current ongoing transaction.